File: engine/internal/personas/totp.go
package personas

import (
	"crypto/hmac"
	"crypto/sha1"
	"encoding/base32"
	"encoding/binary"
	"fmt"
	"net/url"
	"strings"
	"time"
)

// Time based one time passwords, RFC 6238 over RFC 4226.
//
// Implemented here rather than pulled in, for two reasons. It is forty lines
// of standard library, and this repository pins its supply chain deliberately,
// so a dependency for forty lines is a dependency to review, pin, scan and
// eventually update forever. The second reason matters more: both sides of
// this need it. The adapter enrols the secret and the runner produces the code
// from it, and having one implementation means they cannot disagree about
// digits, period or algorithm, which is the usual way a TOTP integration
// fails while every unit test passes.
//
// SHA-1 is the algorithm here, and that is correct rather than an oversight.
// RFC 6238 specifies HMAC-SHA1 as the default, every authenticator app
// implements it, and it is what Supabase, Auth0 and the common libraries
// enrol by default. HMAC-SHA1 is not affected by the collision attacks that
// make SHA-1 unsuitable for signatures.

// TOTPPeriod is the window length every common implementation uses.
const TOTPPeriod = 30 * time.Second

// TOTPDigits is the code length every common implementation uses.
const TOTPDigits = 6

// TOTPCode returns the code for a secret at a moment.
//
// The secret is base32 without padding, which is the form enrolment produces
// and the form an otpauth URL carries.
func TOTPCode(secret string, at time.Time) (string, error) {
	key, err := decodeTOTPSecret(secret)
	if err != nil {
		return "", err
	}
	counter := uint64(at.UTC().Unix()) / uint64(TOTPPeriod.Seconds())
	return hotp(key, counter), nil
}

// TOTPCodeAt returns the code for a counter, which is what a test that has to
// pin down a window uses.
func TOTPCodeAt(secret string, counter uint64) (string, error) {
	key, err := decodeTOTPSecret(secret)
	if err != nil {
		return "", err
	}
	return hotp(key, counter), nil
}

// TOTPValid reports whether a code is the one for a moment.
//
// A one window tolerance either side, which is what a server does, because a
// code typed at the end of a window arrives in the next one. Without it a
// test that runs at the wrong instant fails one time in thirty and gets called
// flaky rather than fixed.
func TOTPValid(secret, code string, at time.Time) bool {
	key, err := decodeTOTPSecret(secret)
	if err != nil {
		return false
	}
	counter := uint64(at.UTC().Unix()) / uint64(TOTPPeriod.Seconds())
	for _, c := range []uint64{counter - 1, counter, counter + 1} {
		if hmac.Equal([]byte(hotp(key, c)), []byte(code)) {
			return true
		}
	}
	return false
}

// hotp is RFC 4226: HMAC the counter, take the dynamic truncation the RFC
// specifies, and reduce it to the digit count.
func hotp(key []byte, counter uint64) string {
	var buf [8]byte
	binary.BigEndian.PutUint64(buf[:], counter)

	mac := hmac.New(sha1.New, key)
	mac.Write(buf[:])
	sum := mac.Sum(nil)

	// The low four bits of the last byte choose where to read from, which is
	// what makes the truncation dynamic rather than a fixed prefix.
	offset := sum[len(sum)-1] & 0x0f
	value := binary.BigEndian.Uint32(sum[offset:offset+4]) & 0x7fffffff

	mod := uint32(1)
	for i := 0; i < TOTPDigits; i++ {
		mod *= 10
	}
	return fmt.Sprintf("%0*d", TOTPDigits, value%mod)
}

// decodeTOTPSecret accepts the secret in the forms it is written in.
//
// Authenticator apps show the secret in spaced groups and some systems store
// it padded, so both are accepted rather than rejected as malformed. Anything
// that is not base32 at all is an error, because a silently wrong key produces
// codes that are wrong forever and look like a broken application.
func decodeTOTPSecret(secret string) ([]byte, error) {
	cleaned := strings.ToUpper(strings.NewReplacer(" ", "", "-", "").Replace(secret))
	cleaned = strings.TrimRight(cleaned, "=")
	if cleaned == "" {
		return nil, fmt.Errorf("the TOTP secret is empty")
	}
	key, err := base32.StdEncoding.WithPadding(base32.NoPadding).DecodeString(cleaned)
	if err != nil {
		return nil, fmt.Errorf("the TOTP secret is not base32: %w", err)
	}
	return key, nil
}

// TOTPURI returns the otpauth URL for a secret.
//
// Written into the persona metadata so that a person debugging a run can add
// the same factor to their own authenticator and see what the agent sees.
// Every parameter is stated rather than left to a default, because the
// defaults are what differ between implementations.
func TOTPURI(issuer, account, secret string) string {
	q := url.Values{}
	q.Set("secret", secret)
	q.Set("issuer", issuer)
	q.Set("algorithm", "SHA1")
	q.Set("digits", fmt.Sprintf("%d", TOTPDigits))
	q.Set("period", fmt.Sprintf("%d", int(TOTPPeriod.Seconds())))
	return "otpauth://totp/" + url.PathEscape(issuer+":"+account) + "?" + q.Encode()
}

File: runner/src/totp.ts
// Time based one time passwords, RFC 6238 over RFC 4226.
//
// The other half of engine/internal/personas/totp.go. The engine enrols the
// secret and this produces the code from it, so the two have to agree about
// digits, period and algorithm. They are written from the same RFC and both
// are checked against the same Appendix B vectors, which is what makes
// "they agree" a fact rather than a hope: a TOTP integration usually fails at
// exactly this seam while every unit test on either side passes.
//
// SHA-1 is correct here rather than an oversight. RFC 6238 specifies
// HMAC-SHA1 as the default and every authenticator app implements it. The
// collision attacks that make SHA-1 unsuitable for signatures do not apply to
// HMAC-SHA1.

import { createHmac } from 'node:crypto';

/** The window length every common implementation uses, in seconds. */
export const PERIOD = 30;

/** The code length every common implementation uses. */
export const DIGITS = 6;

/** decodeSecret accepts the secret in the forms it is written in.
 *
 * Authenticator apps show it in spaced groups and some systems store it
 * padded, so both are accepted. Anything that is not base32 throws, because a
 * silently wrong key produces codes that are wrong forever and look exactly
 * like an application refusing a correct one.
 */
export function decodeSecret(secret: string): Buffer {
  const cleaned = secret.replace(/[\s-]/g, '').toUpperCase().replace(/=+$/, '');
  if (cleaned === '') throw new Error('The TOTP secret is empty.');

  const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
  let bits = 0;
  let value = 0;
  const out: number[] = [];
  for (const char of cleaned) {
    const index = alphabet.indexOf(char);
    if (index === -1) throw new Error(`The TOTP secret is not base32: ${char} is not a base32 character.`);
    value = (value << 5) | index;
    bits += 5;
    if (bits >= 8) {
      bits -= 8;
      out.push((value >>> bits) & 0xff);
    }
  }
  return Buffer.from(out);
}

/** hotp is RFC 4226: HMAC the counter, take the dynamic truncation, reduce. */
function hotp(key: Buffer, counter: number): string {
  const buf = Buffer.alloc(8);
  // Written as two 32 bit halves because a JavaScript number cannot hold a
  // 64 bit integer exactly, and writeBigUInt64BE would need a BigInt for a
  // value that is always small.
  buf.writeUInt32BE(Math.floor(counter / 2 ** 32), 0);
  buf.writeUInt32BE(counter >>> 0, 4);

  const sum = createHmac('sha1', key).update(buf).digest();

  // The low four bits of the last byte choose where to read from, which is
  // what makes the truncation dynamic rather than a fixed prefix.
  const offset = sum[sum.length - 1]! & 0x0f;
  const value =
    ((sum[offset]! & 0x7f) << 24) |
    (sum[offset + 1]! << 16) |
    (sum[offset + 2]! << 8) |
    sum[offset + 3]!;

  return String(value % 10 ** DIGITS).padStart(DIGITS, '0');
}

/** totpCode returns the code for a secret at a moment, in milliseconds. */
export function totpCode(secret: string, atMs: number = Date.now()): string {
  const counter = Math.floor(atMs / 1000 / PERIOD);
  return hotp(decodeSecret(secret), counter);
}

/** secondsIntoWindow says how far through the current window a moment is.
 *
 * Used to decide whether to wait. A code produced with two seconds left is one
 * the application may well reject by the time it is typed, and waiting three
 * seconds is much cheaper than a failed sign in that reads as a real bug.
 */
export function secondsIntoWindow(atMs: number = Date.now()): number {
  return Math.floor(atMs / 1000) % PERIOD;
}
